### PR TITLE
[Validator] add a steps options to the Range Validator

### PR DESCRIPTION
**Pull request by Simperfit**
| Q             | A
| ------------- | ---
| Branch?       | 4.1
| Bug fix?      | no
| New feature?  | yes 
| BC breaks?    | no
| Deprecations? | no 
| Tests pass?   | yes
| Fixed tickets | #25982  
| License       | MIT
| Doc PR        | todo when PR validated.

This PR adds a way to valid a range by steps.

### CODE REVIEW

File: src/Symfony/Component/Validator/Constraints/RangeValidator.php [hunk 1]
```diff
@@ -76,5 +76,15 @@ public function validate($value, Constraint $constraint)
                 ->setCode(Range::TOO_LOW_ERROR)
                 ->addViolation();
         }
+
+        if (null !== ($step = $constraint->step) && !\in_array($value, range($min, $max, $step), true)) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Validator/Constraints/RangeValidator.php]: Shouldn't the last argument of `in_array()` be false? I feel like this should handle stringy numbers, since this is what forms always get as input. (and the rest of the validator works with them already)
Please add a few test cases with strings as values (numbers and not numbers, same for min/max tests).

**Review by nicolas-grekas — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Validator/Constraints/Range.php [hunk 2]
```diff
@@ -25,18 +25,22 @@ class Range extends Constraint
...
     protected static $errorNames = array(
         self::INVALID_CHARACTERS_ERROR => 'INVALID_CHARACTERS_ERROR',
         self::TOO_HIGH_ERROR => 'TOO_HIGH_ERROR',
         self::TOO_LOW_ERROR => 'TOO_LOW_ERROR',
+        self::INVALID_STEP_ERROR => 'INVALID_STEP_ERROR',
     );
 
     public $minMessage = 'This value should be {{ limit }} or more.';
     public $maxMessage = 'This value should be {{ limit }} or less.';
+    public $stepMessage = 'This value should be between {{ min }} and {{ max }} with a step of {{ step }}';
```
Comment (Destroy666x) [inline src/Symfony/Component/Validator/Constraints/Range.php]: What about adding it to English translation file? New to this repository, not sure if that's done in separate PRs.

File: src/Symfony/Component/Validator/Constraints/RangeValidator.php (hunk 1 above)
Comment (Simperfit) [reply src/Symfony/Component/Validator/Constraints/RangeValidator.php]: i'm doing it.

**Comment by Simperfit** [comment]
@Destroy666x I guess that could be done, do we want to be done in this PR ?

File: src/Symfony/Component/Validator/Constraints/Range.php (hunk 2 above)
Comment (Simperfit) [reply src/Symfony/Component/Validator/Constraints/Range.php]: I don't know, @xabbuh WDYT ?

Comment (dunglas) [reply src/Symfony/Component/Validator/Constraints/Range.php]: I think you can do it directly in this PR.

Comment (Simperfit) [reply src/Symfony/Component/Validator/Constraints/Range.php]: Doing it.

Comment (Simperfit) [reply src/Symfony/Component/Validator/Constraints/Range.php]: translation added

**Comment by Simperfit** [comment]
Status: Needs Review

File: src/Symfony/Component/Validator/Resources/translations/validators.pl.xlf [hunk 3]
```diff
@@ -310,9 +310,9 @@
                 <source>This value does not match the expected {{ charset }} charset.</source>
                 <target>Ta wartość nie pasuje do oczekiwanego zestawu znaków {{ charset }}.</target>
             </trans-unit>
-            <trans-unit id="81">
-                <source>This is not a valid Business Identifier Code (BIC).</source>
-                <target>Ta wartość nie jest poprawnym kodem BIC (Business Identifier Code).</target>
+            <trans-unit id="82">
```
Comment (Destroy666x) [inline src/Symfony/Component/Validator/Resources/translations/validators.pl.xlf]: Removed 81 by mistake

Comment (Simperfit) [reply src/Symfony/Component/Validator/Resources/translations/validators.pl.xlf]: Thanks, fixed :).

File: src/Symfony/Component/Validator/Constraints/RangeValidator.php [hunk 4]
```diff
@@ -76,5 +76,15 @@ public function validate($value, Constraint $constraint)
                 ->setCode(Range::TOO_LOW_ERROR)
                 ->addViolation();
```
Comment (Destroy666x) [inline src/Symfony/Component/Validator/Constraints/RangeValidator.php]: One more thing, should a `return` be added here? If value is below `min` you'll get two errors now, for going above `max` only one. The error for range should be enough.

Comment (Simperfit) [reply src/Symfony/Component/Validator/Constraints/RangeValidator.php]: You are right, thanks !

**Comment by Simperfit** [comment]
fabbot error is a false positive.

File: src/Symfony/Component/Validator/Constraints/Range.php [hunk 5]
```diff
@@ -25,18 +25,22 @@ class Range extends Constraint
     const INVALID_CHARACTERS_ERROR = 'ad9a9798-7a99-4df7-8ce9-46e416a1e60b';
     const TOO_HIGH_ERROR = '2d28afcb-e32e-45fb-a815-01c431a86a69';
     const TOO_LOW_ERROR = '76454e69-502c-46c5-9643-f447d837c4d5';
+    const INVALID_STEP_ERROR = 'de5ca5b7-8fcf-49cf-97f7-1319b4fe8e5e';
```
Comment (xabbuh) [inline src/Symfony/Component/Validator/Constraints/Range.php]: `public const`?

File: src/Symfony/Component/Validator/Constraints/RangeValidator.php [hunk 6]
```diff
@@ -75,6 +75,18 @@ public function validate($value, Constraint $constraint)
                 ->setParameter('{{ limit }}', $this->formatValue($min, self::PRETTY_DATE))
                 ->setCode(Range::TOO_LOW_ERROR)
                 ->addViolation();
+
+            return;
+        }
+
+        if (null !== ($step = $constraint->step) && !\in_array($value, range($min, $max, $step), false)) {
```
Comment (xabbuh) [inline src/Symfony/Component/Validator/Constraints/RangeValidator.php]: How will that work with dates?

Comment (Simperfit) [reply src/Symfony/Component/Validator/Constraints/RangeValidator.php]: It does not work with date, that was not the feature, but we can do it like @Destroy666x said with a dateTimeRange.

File: src/Symfony/Component/Validator/Constraints/Range.php (hunk 5 above)
Comment (Simperfit) [reply src/Symfony/Component/Validator/Constraints/Range.php]: isn't it public by default ?

Comment (xabbuh) [reply src/Symfony/Component/Validator/Constraints/Range.php]: Yes, but I think we should be explicit now that we can do that for constants (just like we deal with class properties).

File: src/Symfony/Component/Validator/Constraints/RangeValidator.php (hunk 6 above)
Comment (xabbuh) [reply src/Symfony/Component/Validator/Constraints/RangeValidator.php]: Well, I think we at least need to provide a solution that does not break when dates are passed.

Comment (Simperfit) [reply src/Symfony/Component/Validator/Constraints/RangeValidator.php]: I guess we can do this and then provide the feature after if we need it ;).

**Comment by Destroy666x** [comment]
What about the case I mentioned in "edit" here: https://github.com/symfony/symfony/pull/26199#issuecomment-366677409? With `min = 10`, `max = null` and `step=1` you'll get numbers from 10 to 0. Checking whether something is divisible rather than using `range()` would work.

But now that I'm looking at it, why not force both `min` and `max` for Range (apart from BC)? There are validators like `GreaterThanOrEqual` that should be enough for a single limitation.

**Comment by Simperfit** [comment]
We can force min and max for range, there will be not BC since this is a new feature, I think that's the best way to deal with it.

File: src/Symfony/Component/Validator/Constraints/RangeValidator.php (hunk 6 above)
Comment (ostrolucky) [reply src/Symfony/Component/Validator/Constraints/RangeValidator.php]: I think this should be reverted. Solution SHOULD break if DateTime is passed here and `step` is defined. Otherwise it silently works, which makes harder debugging.

It won't break existing code, since this is evaluated only if step is not null.

**Review by ostrolucky — CHANGES_REQUESTED** [review]
Yep, dealing with unbounded min/max is missing here

**Comment by Simperfit** [comment]
done

**Comment by Destroy666x** [comment]
Coming back to this, anyone disagrees that both `min` and `max` should be both required considering that there are other validators that cover the rest and the functionality of e.g. `$min = null` is basically a duplicate that forces additional conditions in features like this?

**Comment by apfelbox** [comment]
As for the implementation + initial request in #25982

Shouldn't this now be implemented using a combination of `Range` with `DivisibleBy`?

**Comment by ostrolucky** [comment]
indeed

**Comment by fabpot** [comment]
Closing then.
